### DISCUSSION TITLE
Is there built-in DelegateCommand for binding command in style setter?

### DISCUSSION THREAD

**Discussion by GF-Huang** [canton7/Stylet#202] [Q&A] [answered]
I had read:

> Actions will not work from style setters. The classes required to do this in WPF are all internal, which means there is no way to fix the issue. You will need to use old-fashioned Commands in this (rare) case, unfortunately.

**Comment by canton7** [comment]
That documentation is still correct. Stylet does not provide an ICommand implementation (if that's what you're referring to with DelegateCommand), but they're pretty trivial to write

**Reply by GF-Huang** [reply]
Why not built-in one?

**Comment by canton7** [comment, marked as answer]
There isn't much point, and it would take someone longer to discover and understand Stylet's version than to just write their own.

**Reply by GF-Huang** [reply]
So everybody who use Stylet with style setter command binding (generally in `MenuItem.ItemContainerStyle`) need to write a `DelegateCommand` again. 😃

**Reply by canton7** [reply]
Yep, you got it!

There are at least 2 different schools of thought on ICommand implementations, and 95% of users won't need one anyway. I don't want to push a particular implementation on people, and I don't want people to discover the ICommand implementation before they discover actions, and so never use actions.

You could have written an ICommand implementation in the time you've taken complaining about it here 😄 In fact, here's one for you now:

```cs
public class RelayCommand : ICommand
{
    private readonly Action<object> action;

    private bool _canExecute;
    public bool CanExecute
    {
        get => this._canExecute;
        set
        {
            if (value == this._canExecute)
                return;
            this._canExecute = value;
            this._canExecuteChanged?.Invoke(this, EventArgs.Empty);
        }
    }

    public RelayCommand(Action<object> action) => this.action = action;

    private event EventHandler _canExecuteChanged;
    event EventHandler ICommand.CanExecuteChanged
    {
        add => this._canExecuteChanged += value;
        remove => this._canExecuteChanged -= value;
    }
    bool ICommand.CanExecute(object parameter) => this.CanExecute;
    void ICommand.Execute(object parameter) => this.action?.Invoke(parameter);
}
```

Feel free to use that in all your projects!

**Reply by GF-Huang** [reply]
Why need a explicit implementation for `ICommand`?

**Reply by canton7** [reply]
The ICommand members are only for use by the UI, not for your VM to call.

This is what I mean when I say that everyone has their favourite way to implement it. Use an implementation which works for you
